### PR TITLE
Add com.openwall.John (JtR Jumbo)

### PR DESCRIPTION
**Pull request by claudioandre-br**
The package builds a fake release:
- You can review first and I can announce the FlatHub package later, during release(?);
- In fact, I would like to reference FlatHub [1] inside the appdata.xml file. So, I need to see how the review will happen.
- Also, it is not clear to me how many archs Flathub has.

See also
- a (testing) flatpak bundle: https://gitlab.com/claudioandre-br/packages/-/jobs/194731622/artifacts/browse
- John as snap "package": https://snapcraft.io/john-the-ripper

[1] https://flathub.org/apps/detais/com.openwall.John

### CODE REVIEW

File: use-regex-1.4.patch [hunk 1]
```diff
@@ -0,0 +1,164 @@
+From 0509d49c4bd5931a78b4f3df5771cdd2d369e238 Mon Sep 17 00:00:00 2001
```
Comment (bilelmoussaoui) [inline use-regex-1.4.patch]: Nitpick: You can put those patches in a directory to keep the repository clean once created.

File: flathub.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "skip-appstream-check": true
```
Comment (bilelmoussaoui) [inline flathub.json]: This should be removed. We need to ensure that the appdata is valid and correctly installed.

File: com.openwall.John.json [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                    "type": "file",
+                    "path": "com.openwall.John.png"
+                }
+            ],
+            "post-install": [
+                "###### Install JtR system wide ######",
+                "cp -r ../run/.   /app/bin",
+                "cp -r ../doc/    /app/",
+                "cp    ../README* /app/",
+                "###### Bash Completion ######",
```
Comment (bilelmoussaoui) [inline com.openwall.John.json]: bash completion is not exported. Not sure if it's useful

File: com.openwall.John.json [hunk 4]
```diff
@@ -0,0 +1,94 @@
+{
+    "id": "com.openwall.John",
+    "runtime": "org.freedesktop.Platform",
+    "runtime-version": "1.6",
```
Comment (bilelmoussaoui) [inline com.openwall.John.json]: Use 18.08 instead.

File: com.openwall.John.json [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    "sdk": "org.freedesktop.Sdk",
+    "command": "john",
+    "tags": [ "1.9J1-4bf9652", "1.9.0 Jumbo 1" ],
```
Comment (bilelmoussaoui) [inline com.openwall.John.json]: Tags are not used here on flathub.

File: com.openwall.John.json [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    "finish-args": [
+        "--filesystem=home"
```
Comment (bilelmoussaoui) [inline com.openwall.John.json]: does the app only requires home access, no gui or anything else?

File: com.openwall.John.desktop [hunk 7]
```diff
@@ -0,0 +1,12 @@
+[Desktop Entry]
+Version=1.9.0J1
+Type=Application
+Terminal=1
+Name=John the Ripper Community Edition
+GenericName=John the Ripper Jumbo
+Comment=John the Ripper "Jumbo" password cracker
+TryExec=john
```
Comment (bilelmoussaoui) [inline com.openwall.John.desktop]: Should be upstream.

File: com.openwall.John.appdata.xml [hunk 8]
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Copyright (c) 2019 Claudio André <claudioandre.br at gmail.com> -->
```
Comment (bilelmoussaoui) [inline com.openwall.John.appdata.xml]: The appdata file should be upstream.

File: com.openwall.John.appdata.xml [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+       Read more at:
+       - github.com/magnumripper/JohnTheRipper
+       - rebrand.ly/JtRPackage
+
+       This version integrates lots of contributed patches including GPU support.
+    </p>
+  </description>
+
+  <icon type="local">/app/share/icons/com.openwall.John.png</icon>
```
Comment (bilelmoussaoui) [inline com.openwall.John.appdata.xml]: Not needed. 
```suggestion
```

File: com.openwall.John.appdata.xml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+          alone is still sufficient or where (cross-)compiling jumbo for a given
+          target system is too difficult or (as a first step in) porting John the
+          Ripper to an unusual new platform.
+        </p>
+      </description>
+
+      <url>https://github.com/claudioandre-br/JohnTheRipper/releases/tag/1.9.0-jumbo-1</url>
+
+      <artifacts>
+        <artifact type="binary" platform="x86_64-linux-gnu">
```
Comment (bilelmoussaoui) [inline com.openwall.John.appdata.xml]: Not sure if it's a good idea to have this here.

File: com.openwall.John.appdata.xml [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  </screenshots>
+
+  <url type="homepage">https://www.openwall.com/john/</url>
+  <url type="bugtracker">https://github.com/magnumripper/JohnTheRipper/issues</url>
+  <url type="donation">https://www.zerodayclothing.com/openwall_store.php</url>
+
+  <developer_name>Openwall and others</developer_name>
+  <update_contact>claudioandre.br@gmail.com</update_contact>
+
+  <categories>
```
Comment (bilelmoussaoui) [inline com.openwall.John.appdata.xml]: Categories should be placed only on the desktop file, they are later merged using appstream-compose.

File: build.sh [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+# Copyright (c) 2019 Claudio André <claudioandre.br at gmail.com>
+#
+# This program comes with ABSOLUTELY NO WARRANTY; express or implied.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, as expressed in version 2, seen at
+# http://www.gnu.org/licenses/gpl-2.0.html
+######################################################################
+
```
Comment (bilelmoussaoui) [inline build.sh]: This should be upstream, not sure if it's something flatpak specific or you can not achieve with a "real" build system.

**Comment by bilelmoussaoui** [comment]
you can test the build locally which is easier for you:
```
flatpak-builder --install app-repo com.openwall.John.json --user
```

**Comment by LeandroStanger** [comment]
bot, build com.openwall.John

File: build.sh (hunk 12 above)
Comment (claudioandre-br) [reply build.sh]: This is packaging specific. A real user builds the binary he/she is going to use. A package should have SSE2, AVX, .... And them (the binaries packaged) have to fallback.

File: com.openwall.John.json (hunk 6 above)
Comment (claudioandre-br) [reply com.openwall.John.json]: Yes.

File: com.openwall.John.json (hunk 3 above)
Comment (claudioandre-br) [reply com.openwall.John.json]: The examples I found suggest that copy to the _magic place_ will work. Do you have any example that I can follow?

File: com.openwall.John.json (hunk 5 above)
Comment (claudioandre-br) [reply com.openwall.John.json]: I like them. Should I remove?

Comment (bilelmoussaoui) [reply com.openwall.John.json]: Well, they are not used anyway. So they should be removed here

File: build.sh [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        make -s clean && make -sj4
+    fi
+    set +e
+}
+
+#if (Build); then
+    echo ""
+    echo "---------------------------- BUILDING -----------------------------"
+
+    if [[ "$(uname -m)" == "x86_64" || "$(uname -m)" == "i386" || "$(uname -m)" == "i686" ]]; then
```
Comment (bilelmoussaoui) [inline build.sh]: You can achieve that with config-opts.

Comment (claudioandre-br) [reply build.sh]: If you don't mind, I would like to keep this as is (using a shell script):
* I have to set the build options (ok). But;
  * I have to rename the binary;
  * I have to inform the next binary the name of the previous one, escaping whatever needed.
  * Some binaries fallback to OpenMP, some to an extension plus OpenMP;

Plus, that I'm sharing the script with other build systems: snap and Windows. Well, and I saw no `config-opts` use-case complex as this one.

**Comment by LeandroStanger** [comment]
bot, build com.openwall.John

**Comment by claudioandre-br** [comment]
* The "packaging" files were merged upstream.
* Nothing new for a while, please ping me if you need anything else.

Sorry, BTW. I missed one change in the previous patch.

Please, let me know if I have to do something about the message: 
* `org.freedesktop.Sdk/x86_64/18.08 not installed`

**Comment by LeandroStanger** [comment]
bot, build com.openwall.John

**Comment by nedrichards** [comment]
@claudioandre-br flathub supports x86-64, x86, arm and aarch64 architectures.

**Comment by nedrichards** [comment]
> This version integrates lots of contributed patches, including GPU support,

It won't have access to a GPU unless you add --device=dri

**Comment by claudioandre-br** [comment]
> flathub supports x86-64, x86, arm and aarch64 architectures.

Thanks. I had noticed reading the build bot logs. BTW: when I see arm, I read armhf.

> It won't have access to a GPU unless you add --device=dri

The (proposed) flatpak can't access the GPU (since the `--disable-opencl`). Jumbo itself can. I'm not sure if we will have time and hardware to test the flatpak GPU "mode". So, the roadmap: flatpak is/will be CPU only.

**Comment by claudioandre-br** [comment]
>It won't have access to a GPU unless you add --device=dri

BTW: I just noticed 18.08 SDK does not have OpenCL (1.6 has). So, no OpenCL for now. Anyway, thanks for the advice.

**Comment by nedrichards** [comment]
@claudioandre-br in that case can you change the appdata description to correctly reflect the options that will be available. Not an issue whith whatever configuration you choose, just trying to remove a potential point of user confusion.

**Comment by claudioandre-br** [comment]
I changed appdata text. BTW:
* OpenCL will not work properly. I tested using 1.6, but I bet 18.08 is no different. 
* OpenCL proprietary drivers are installed on **host**.
* And I (the OpenCL library as stated in the OpenCL standard) need to access them.
* Links to host file system are missing. Plus, the capacity to load a library from host.

What I did:
```
$ flatpak run --filesystem=host --command=sh --devel com.openwall.John

# OpenCL drivers are installed on **host**
$ rm /etc/OpenCL
$ ln -s /run/host/etc/OpenCL/ /etc/OpenCL # maybe flatpak has a way to specify this

# ICDs will load the driver installed on **host**
$ export LD_LIBRARY_PATH=/usr/lib:/run/host/usr/lib64/

# Ops, I need to do this (for POCL provider). But the folder is RO. Cant test to the end.
ln -s /run/host/usr/share/pocl /usr/share/pocl
```

**Comment by LeandroStanger** [comment]
bot, build com.openwall.John

**Comment by nedrichards** [comment]
@claudioandre-br there's been some OpenCL work (basically making sure things are exported correctly for the flatpack container to see it) which you can look at here https://github.com/flatpak/flatpak/issues/574 but it likely needs verfification and further work.

**Comment by claudioandre-br** [comment]
Thanks. But the release is expected to be in a week or so. For us is safe to keep it a CPU app.

File: build.sh (hunk 13 above)
Comment (barthalion) [reply build.sh]: Looks like that's what more or less other distributions are doing so I guess we will have to live with it.

File: com.openwall.John.desktop (hunk 7 above)
Comment (barthalion) [reply com.openwall.John.desktop]: I'm not actually if it's good idea to provide desktop launcher if application doesn't provide TUI.

**Comment by nedrichards** [comment]
@claudioandre-br let me know when you're ready to release and I'll happily merge. Is there anyone else who needs to be added to the repo to manage it?

**Comment by claudioandre-br** [comment]
Em ter, 30 de abr de 2019 às 05:58, Nick Richards <notifications@github.com>
escreveu:

> @claudioandre-br <https://github.com/claudioandre-br> let me know when
> you're ready to release and I'll happily merge. Is there anyone else who
> needs to be added to the repo to manage it?
>

Thanks. Still waiting the release!

**Comment by claudioandre-br** [comment]
Software released and ready to deploy. Thank you all guys for your hard work.

**Comment by LeandroStanger** [comment]
bot, build com.openwall.John

**Comment by claudioandre-br** [comment]
It is complaining (now) about OARS.
- I computed one at https://hughsie.github.io/oars/generate.html and added it to JSON.

**Comment by claudioandre-br** [comment]
Since it is passing the test below, and it was working recently (Build 2666 successful), I'm puzzled.

```
$ appstream-util validate-relax com.openwall.John.appdata.xml
com.openwall.John.appdata.xml: OK
```

**Comment by barthalion** [comment]
bot, build com.openwall.John

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/com.openwall.John

**Pull request closed without merging**